src/xen_bus.c

```diff
@@ -0,0 +1,771 @@
+			sys_dlist_append(inspected_list, &iter->node);
+		}
+
+		inspected_list = &iter->childs;
+
+		if (path[slashpos] == 0) {
+			if (vall > 0) {
+				if (iter->value != NULL) {
+					k_free(iter->value);
+				}
+
+				iter->value = k_malloc(vall);
+				memcpy(iter->value, data, vall);
+			}
+
+			break;
+		}
+
+		slashpos_prev = slashpos;
+	}
+
+	k_mutex_unlock(&xsel_mutex);
+}
+
+void _handle_write(struct xen_domain *domain, uint32_t id, uint32_t msg_type, char *payload,
+		   uint32_t len)
+{
+	uint32_t data_offset = strlen(payload) + 1;
+	char *data = payload + data_offset;
+
+	if (len < data_offset) {
+		printk("SIZES MISMATCH\n");
+		send_errno(domain, id, EINVAL);
+		return;
+	}
+
+	const char localpath[] = "/";
+	char path[128];
+
+	if (memcmp(payload, localpath, strlen(localpath)) == 0) {
+		memcpy(path, payload, data_offset);
+	} else {
+		snprintf(path, 128, "/local/domain/%d/%s", domain->domid, payload);
+	}
+
+	data[len - data_offset] = 0;
+	do_write(path, data);
+
+	send_reply(domain, id, msg_type, "OK");
+
+	struct watch_entry *iter;
+
+	k_mutex_lock(&wel_mutex, K_FOREVER);
+	SYS_DLIST_FOR_EACH_CONTAINER (&watch_entry_list, iter, node) {
+		uint32_t iklen = strlen(iter->key);
+		if (memcmp(iter->key, path, iklen) == 0) {
+			struct pending_watch_event_entry *entry;
+			entry = k_malloc(sizeof(*entry));
+			int keyl = strlen(path) + 1;
+			entry->key = k_malloc(keyl);
+			memcpy(entry->key, path, keyl);
+			entry->key[keyl - 1] = 0;
+			entry->domid = iter->domid;
+			k_mutex_lock(&pfl_mutex, K_FOREVER);
+			sys_dnode_init(&entry->node);
+			sys_dlist_append(&pending_watch_event_list, &entry->node);
+			k_mutex_unlock(&pfl_mutex);
+		}
+	}
+
+	k_mutex_unlock(&wel_mutex);
+}
+
+void handle_write(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	_handle_write(domain, id, XS_WRITE, payload, len);
+}
+
+void handle_mkdir(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	_handle_write(domain, id, XS_MKDIR, payload, len);
+}
+
+void process_pending_watch_events(struct xen_domain *domain, uint32_t id)
+{
+	struct pending_watch_event_entry *iter, *next;
+	k_mutex_lock(&pfl_mutex, K_FOREVER);
+
+	SYS_DLIST_FOR_EACH_CONTAINER_SAFE (&pending_watch_event_list, iter, next, node) {
+		if (domain->domid == iter->domid) {
+			if (fire_watcher(domain, id, iter->key)) {
+				k_free(iter->key);
+				sys_dlist_remove(&iter->node);
+				k_free(iter);
+				k_mutex_unlock(&pfl_mutex);
+			}
+		}
+	}
+
+	k_mutex_unlock(&pfl_mutex);
+}
+
+void handle_control(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	send_reply(domain, id, XS_CONTROL, "OK");
+}
+
+void handle_get_perms(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	send_errno(domain, id, ENOSYS);
+}
+
+void handle_set_perms(struct xen_domain *domain, uint32_t id, char *payload, uint32_t len)
+{
+	send_reply(domain, id, XS_SET_PERMS, "OK");
+}
+
+void remove_watch_entry(struct watch_entry *entry)
```

Comment:
As I can see from usage - this should be "static" or explicitly told, that "should be called with `wel_mutex` taken"